Spark SQL based Shark CLI

liancheng opened the pull request:
Merged PR #337 authored by @chenghao-intel, and then did some refactoring and update, mainly include:
- Minimized `CatalystContext` as we don't care about response code for now, should update after [SPARK-2106](https://issues.apache.org/jira/browse/SPARK-2106) is addressed.
- `CatalystDriver.destroy` should call `super.destroy`.
- A bunch of trivial coding style fix (including original Shark code).

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/amplab/shark/341)

<!-- Reviewable:end -->

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished.

AmplabJenkins commented:
Refer to this link for build results: https://amplab.cs.berkeley.edu/jenkins/job/Shark-Pull-Request-Builder/12211/

rxin commented on src/main/scala/org/apache/spark/sql/hive/CatalystContext.scala:
```diff
@@ -0,0 +1,32 @@
+
+import shark.LogHelper
+
+class CatalystContext(sc: SparkContext) extends HiveContext(sc) with LogHelper {
```
SparkSQLContext?

marmbrus replied:
or JDBCContext?

liancheng replied:
Hmm... I'd rather just call it `SharkContext`. Catalyst is a query optimization framework, Spark SQL is more than that, but neither of them is a concept parallel to Hive. And this class is actually not very much related to JDBC.

marmbrus replied:
SGTM

rxin commented on src/main/scala/shark/CatalystDriver.scala:
```diff
@@ -0,0 +1,89 @@
+  override def getSchema: Schema = tableSchema
+
+  override def getResults(res: JArrayList[String]): Boolean = {
+    if(hiveResponse == null) {
```
space after if

rxin commented on src/main/scala/shark/CatalystEnv.scala:
```diff
@@ -0,0 +1,150 @@
+  //com.esotericsoftware.minlog.Log.set(com.esotericsoftware.minlog.Log.LEVEL_DEBUG)
+
+  // Keeps track of added JARs and files so that we don't add them twice in consecutive queries.
+  val addedFiles = HashSet[String]()
```
are these two still useful?

rxin commented on src/main/scala/shark/SharkCliDriver.scala:
```diff
@@ -220,24 +224,32 @@ object SharkCliDriver {
         prefix = ""
-        val sharkMode = SharkConfVars.getVar(conf, SharkConfVars.EXEC_MODE) == "shark"
-        curPrompt = if (sharkMode) SharkCliDriver.prompt else CliDriver.prompt
+        val sharkMode = SharkConfVars.getVar(conf, SharkConfVars.EXEC_MODE) == "catalyst"
```
i think we can get rid of the alternative hive exec mode now.

chenghao-intel replied:
That may be helpful if compare the result between Hive & Catalyst, or can we just NOT document it and only for internal usage?

rxin replied:
you can always download hive to use, no?

rxin replied:
either way we can have it, but i am not sure if a single variable justifies a separate config system.

chenghao-intel replied:
1. Yes, I think the downloaded Hive can be used here as official released Hive, otherwise it should be a bug.
2. Yep, you're right, we'd better get rid of SharkConfVars.scala, as we have the plan to copy all of the configuration key names to spark sql conf.

rxin commented on src/main/scala/shark/SharkConfVars.scala:
```diff
@@ -0,0 +1,186 @@
+import org.apache.hadoop.hive.conf.HiveConf
+
+
+object SharkConfVars {
```
i don't think we need this anymore. everything should be in spark sql conf.

chenghao-intel replied:
I was thinking if we need to backward-compatible(with to Shark) for the configuration keys name.

rxin replied:
yup we will build it into spark sql to use the right configs. e.g. spark sql should read mapred.reduce.tasks too if the spark sql version of that is not set.

rxin replied:
if you want to give it a try for mapred.reduce.tasks, please go ahead!

chenghao-intel commented on README.md:
````diff
@@ -59,9 +59,19 @@ resultSet.next()
+catalyst> set shark.exec.mode=hive;
+hive>show tables;
+```
+But there is a bug, which require show tables before doing anything else.
````
This should not be a bug anymore. :)

chenghao-intel commented on README.md:
```diff
@@ -59,9 +59,19 @@ resultSet.next()
+
 ## Known Missing Features
-* Shark CLI
-* Restoring cached tables upon restart
```
This is a missing feature.

liancheng replied:
Thanks, the Spark SQL branch is changing somewhat rapidly, and I'd like to update the README file when it's somewhat stabilized :)

chenghao-intel commented on src/main/scala/shark/SharkCliDriver.scala:
```diff
@@ -246,54 +258,57 @@ object SharkCliDriver {
         "Reloading cached RDDs from previous Shark sessions... (use %s flag to skip reloading)"
         .format(SharkCliDriver.SKIP_RDD_RELOAD_FLAG))
-      TableRecovery.reloadRdds(processCmd(_), Some(console))
+//      TableRecovery.reloadRdds(processCmd(_), Some(console), ss)
```
Add TODO here, this is the root where we add back the code of reloading the cached tables.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished.

AmplabJenkins commented:
Refer to this link for build results: https://amplab.cs.berkeley.edu/jenkins/job/Shark-Pull-Request-Builder/12212/

liancheng commented:
@rxin @chenghao-intel, thanks for your comments :) I've removed the alternative Hive execution mode, and addressed issues you brought up. I cleaned up some code that doesn't follow reasonable coding conventions along the way until I realized that most of it was copied from the original Hive codebase. Maybe I should just left them unchanged...
